Improve OIDC multi-tenancy and providers docs (#34659)
quarkusio/quarkus: docs/src/main/asciidoc/security-openid-connect-multitenancy.adoc

In both cases, if the user is successfully authenticated, the landing page will show the user's name and e-mail. Even though
user `alice` exists in both tenants, for the application they are distinct users belonging to different realms/tenants.

[[static-tenant-resolution]]

> == Static Tenant configuration resolution

Reviewer: Sentence case as per Quarkus title style:
```suggestion
== Static tenant configuration resolution
```

If this heading is a required step in this how-to guide to complete the end-to-end solution, use a more [user-focussed and goal-oriented heading](https://quarkus.io/version/main/guides/doc-reference#titles-headings), for example : 

  ```suggestion
== Configure the resolution of static tenants
```

or 

  ```suggestion
== Configure static tenant resolution
```

> If you have prepared multiple tenant configurations in `application.properties` then it is only a tenant identifier
> that has to be resolved and you can use one of following 3 options to resolve tenant identifiers:

Reviewer: ```suggestion
When you set multiple tenant configurations in the `application.properties` file, you only need to specify how the tenant identifier gets resolved.
To configure the resolution of the tenant identifier, use one of the following options:

* <<default-tenant-resolver>>
* <<tenant-resolver>>
* <<annotations-tenant-resolver>>
```

[[default-tenant-resolver]]

> === Default Resolution

Reviewer: ```suggestion
=== Default resolution
```

> Default tenant identifier resolution is convention based, it expects requests requiring authentication to have a tenant identifier set as the last request path segment value.

Reviewer: ```suggestion
The default resolution for a tenant identifier is convention based, whereby the authentication request must include the tenant identifier in the last segment of the request path. 
```

> Lets assume tenant `google` and `github` configurations are set in `application.properties`:

Reviewer: ```suggestion
The following `application.properties` example shows how you can configure two tenants named `google` and `github`:  
```

[source,properties]
----
# Tenant 'google' configuration
quarkus.oidc.google.provider=google
quarkus.oidc.google.client-id=${google-client-id}
quarkus.oidc.google.credentials.secret=${google-client-secret}
quarkus.oidc.google.authentication.redirect-path=/signed-in

# Tenant 'github' configuration
quarkus.oidc.github.provider=google
quarkus.oidc.github.client-id=${github-client-id}
quarkus.oidc.github.credentials.secret=${github-client-secret}
quarkus.oidc.github.authentication.redirect-path=/signed-in
----

> These 2 tenants configure OIDC `web-app` applications which will use an authorization code flow to authenticate users and will depend on session cookies after the authentication has taken place. In both cases, after either Google or GitHub have authenticated the current user, the user is returned to the `/signed-in` area for the already authenticated users - for example, it can be a secured resource path on the JAX-RS endpoint.

Reviewer: ```suggestion
In this example, both tenants configure OIDC `web-app` applications to use an authorization code flow to authenticate users and also require session cookies to get generated after the authentication has taken place. 
After either Google or GitHub authenticates the current user, the user gets returned to the `/signed-in` area for authenticated users, for example, a secured resource path on the JAX-RS endpoint.
```

> Now, the only thing that you will need to make the resolution working by default is to add the following:

Reviewer: ```suggestion
Finally, to complete the default tenant resolution, set the following configuration property:
```

[source,properties]
----
quarkus.http.auth.permission.login.paths=/google,/github
quarkus.http.auth.permission.login.policy=authenticated
----

> and, assuming the endpoint is running on `http://localhost:8080`, provide UI options for users to login to either `http://localhost:8080/google` or `http://localhost:8080/github`, without having to add JAX-RS `/google` or `/github` JAX-RS resource paths. Tenant identifiers will also be recorded in the session cookie names after the authentication has been completed and therefore authenticated users will be able to access the secured application area without secured URLs having to contain either `google` or `github` path values.

Reviewer: Suggestion to reduce sentence length and apply sentence-style:
```suggestion
If the endpoint is running on `http://localhost:8080`, you can also provide UI options for users to log in to either `http://localhost:8080/google` or `http://localhost:8080/github`, without having to add specific`/google` or `/github` JAX-RS resource paths. 
Tenant identifiers are also recorded in the session cookie names after the authentication is completed. 
Therefore, authenticated users can access the secured application area without requiring either the `google` or `github` path values to be included in the secured URL.
```

> Default resolution can also work for Bearer token authentication but it may be less practical in this case since a tenant identifier will always have to be set as the last path segment value.

Reviewer: ```suggestion
Default resolution can also work for Bearer token authentication but it might be less practical in this case because a tenant identifier will always need to be set as the last path segment value.
```

[[tenant-resolver]]
=== Resolve with `TenantResolver`

> Lets assume tenant `a` and `b` configurations are set in `application.properties`:

Reviewer: ```suggestion
The following `application.properties` example shows how you can resolve the tenant identifier of two tenants named `a` and `b` by using the `TenantResolver` method: 
```

[source,properties]
----
# Tenant 'a' configuration
quarkus.oidc.a.auth-server-url=http://localhost:8180/realms/quarkus-a
quarkus.oidc.a.client-id=client-a
quarkus.oidc.a.credentials.secret=client-a-secret

# Tenant 'b' configuration
quarkus.oidc.b.auth-server-url=http://localhost:8180/realms/quarkus-b
quarkus.oidc.b.client-id=client-b
quarkus.oidc.b.credentials.secret=client-b-secret
----

> You can have either `a` or `b` tenant id returned from `quarkus.oidc.TenantResolver`:

Reviewer: ```suggestion
You can return the tenant ID of either `a` or `b` from `quarkus.oidc.TenantResolver`:
```

[source,java]
----
import quarkus.oidc.TenantResolver;

public class CustomTenantResolver implements TenantResolver {

    @Override
    public String resolve(RoutingContext context) {
        String path = context.request().path();
        if (path.endsWith("a")) {
            return "a";
        } else if (path.endsWith("b")) {
            return "b";
        } else {
            // default tenant
            return null;
        }
    }
}
----

> In this example, the value of the last request path segment is a tenant id, but indeed, a more involved tenant identifier resolution logic can be implemented.

Reviewer: ```suggestion
In this example, the value of the last request path segment is a tenant ID, but if required, you can implement a more complex tenant identifier resolution logic.
```

> === Resolve with annotations

Reviewer: ```suggestion
[[annotations-tenant-resolver]]
=== Resolve with annotations
```

OIDC provider.

[[tenant-config-resolver]]

> == Dynamic Tenant configuration resolution

Reviewer: ```suggestion
== Dynamic tenant configuration resolution
```
or as per my comment about the previous section 2 heading:

```suggestion
== Configure the resolution of dynamic tenants
```

The main thing is that both headings are aligned, grammatically.

The `OidcTenantConfig` returned from this method is the same used to parse the `oidc` namespace configuration from the `application.properties`. You can populate it using any of the settings supported by the `quarkus-oidc` extension.

> === Tenant Resolution for OIDC `web-app` applications

Reviewer: ```suggestion
=== Tenant resolution for OIDC `web-app` applications
```

> Simplest option for resolving OIDC `web-app` application configiration is to follow steps described in the <<default-tenant-resolver, Default Resolution>> section.

Reviewer 2: ```suggestion
Simplest option for resolving OIDC `web-app` application configuration is to follow steps described in the <<default-tenant-resolver, Default Resolution>> section.
```

> Simplest option for resolving OIDC `web-app` application configiration is to follow steps described in the <<default-tenant-resolver, Default Resolution>> section.

Reviewer: ```suggestion
The simplest option for resolving OIDC `web-app` application configuration is to follow the steps described in the <<default-tenant-resolver>> section.
```
Anchor change bit not essential, but just flagging that Guillaume wrote a nice script as part of the doc build to convert Naked anchors into an xref that contains the section title text.

Author: @michelle-purcell, George fixed the `configiration` typo with his suggestion earlier, so I could not apply your suggestion directly, thanks

Custom `TenantResolver` and `TenantConfigResolver` implementations may return `null` if no tenant can be inferred from the current request and a fallback to the default tenant configuration is required.

> If it is expected that the custom resolvers will always infer a tenant then the default tenant configuration is not needed. One can disable it with the `quarkus.oidc.tenant-enabled=false` setting but you do not have to do it explicitly - the default tenant will be disabled by default if `quarkus.oidc.auth-server-url` is not configured but either custom tenant configurations are available or `TenantConfigResolver` is registered.

Reviewer: ```suggestion
If you expect that the custom resolvers will always infer a tenant then you do not need to configure the default tenant resolution.

* To disable the default tenant configuration, set  `quarkus.oidc.tenant-enabled=false`.

[NOTE]
====
The default tenant configuration is automatically disabled when `quarkus.oidc.auth-server-url` is not configured but either custom tenant configurations are available or `TenantConfigResolver` is registered.
====
```
